mlir/lib/IR/AffineExpr.cpp

```diff
@@ -237,6 +239,135 @@ unsigned AffineDimExpr::getPosition() const {
+    return binaryExpr.getLHS() * symbolicDivide(binaryExpr.getRHS(), symbolPos);
+  }
+  case AffineExprKind::FloorDiv: {
+    AffineBinaryOpExpr binaryExpr = expr.cast<AffineBinaryOpExpr>();
+    return symbolicDivide(binaryExpr.getLHS(), symbolPos).floorDiv(binaryExpr.getRHS());
+  }
+  case AffineExprKind::CeilDiv: {
+    AffineBinaryOpExpr binaryExpr = expr.cast<AffineBinaryOpExpr>();
+    return symbolicDivide(binaryExpr.getLHS(), symbolPos).ceilDiv(binaryExpr.getRHS());
+  }
+  case AffineExprKind::Mod: {
+    AffineBinaryOpExpr binaryExpr = expr.cast<AffineBinaryOpExpr>();
+    return symbolicDivide(binaryExpr.getLHS(), symbolPos) % binaryExpr.getRHS();
+  }
+  }
+  llvm_unreachable("Unknown AffineExpr");
+}
+
+/// Simplify a semi-affine expression by handling modulo, floordiv or ceildiv
+/// by any symbol.
+static AffineExpr simplifySemi(AffineExpr expr) {
+  switch (expr.getKind()) {
+  case AffineExprKind::Constant:
+    return expr;
+  case AffineExprKind::DimId:
+    return expr;
+  case AffineExprKind::SymbolId:
+    return expr;
+  case AffineExprKind::Add: {
+    AffineBinaryOpExpr binaryExpr = expr.cast<AffineBinaryOpExpr>();
+    return simplifySemi(binaryExpr.getLHS()) + simplifySemi(binaryExpr.getRHS());
+  }
+  case AffineExprKind::Mul: {
+    AffineBinaryOpExpr binaryExpr = expr.cast<AffineBinaryOpExpr>();
+    return simplifySemi(binaryExpr.getLHS()) * simplifySemi(binaryExpr.getRHS());
+  }
+  case AffineExprKind::FloorDiv: {
+    AffineBinaryOpExpr binaryExpr = expr.cast<AffineBinaryOpExpr>();
+    if (isSymbolic(simplifySemi(binaryExpr.getRHS()))) {
+      AffineSymbolExpr sym = binaryExpr.getRHS().dyn_cast<AffineSymbolExpr>();
+      unsigned pos = sym.getPosition();
+      if (isDivisibleBySymbol(binaryExpr.getLHS(), pos)) {
+        return symbolicDivide(simplifySemi(binaryExpr.getLHS()), pos);
+      }
+    }
+    return simplifySemi(binaryExpr.getLHS()).floorDiv(simplifySemi(binaryExpr.getRHS()));
+  }
+  case AffineExprKind::CeilDiv: {
+    AffineBinaryOpExpr binaryExpr = expr.cast<AffineBinaryOpExpr>();
+    if (isSymbolic(simplifySemi(binaryExpr.getRHS()))) {
+      AffineSymbolExpr sym = binaryExpr.getRHS().dyn_cast<AffineSymbolExpr>();
+      unsigned pos = sym.getPosition();
+      if (isDivisibleBySymbol(binaryExpr.getLHS(), pos)) {
+        return symbolicDivide(simplifySemi(binaryExpr.getLHS()), pos);
+      }
+    }
+    return simplifySemi(binaryExpr.getLHS()).ceilDiv(simplifySemi(binaryExpr.getRHS()));
+  }
+  case AffineExprKind::Mod: {
+    AffineBinaryOpExpr binaryExpr = expr.cast<AffineBinaryOpExpr>();
+    if (isSymbolic(simplifySemi(binaryExpr.getRHS()))) {
+      AffineSymbolExpr sym = binaryExpr.getRHS().dyn_cast<AffineSymbolExpr>();
+      unsigned pos = sym.getPosition();
+      if (isDivisibleBySymbol(binaryExpr.getLHS(), pos)) {
+        return getAffineConstantExpr(0, expr.getContext());
+      }
+    }
+    return simplifySemi(binaryExpr.getLHS()) % (simplifySemi(binaryExpr.getRHS()));
```

Comment:
For this and everywhere above, you can use early return for the 'false' case of the conditional - that would reduce the indent for the main part of the code and make it simpler. Eg:

```
if (!isSymbolic ...) 
  return simplifySemi(...);
AffineSymbol...
...
```